File: src/Math/Vector3.ts
import { Vector2 } from "./Vector2";
import { Vector4 } from "./Vector4";

export class Vector3 {
    public x: number;
    public y: number;
    public z: number;

    public constructor();
    public constructor(x: number, y: number, z: number);
    public constructor(v: Vector2);
    public constructor(v: Vector4);
    public constructor() {
        if (arguments[0] instanceof Vector2) {
            this.x = arguments[0].x;
            this.y = arguments[0].y;
            this.z = 0;
        }
        else if (arguments[0] instanceof Vector4) {
            this.x = arguments[0].x;
            this.y = arguments[0].y;
            this.z = arguments[0].z;
        }
        else if (arguments.length == 3) {
            this.x = arguments[0];
            this.y = arguments[1];
            this.z = arguments[2];
        }
        else {
            this.x = this.y = this.z = 0;
        }
    }

    /*
     OPERATIONS ON VECTOR
     */

    public add(v: Vector3): Vector3 {
        this.x += v.x;
        this.y += v.y;
        this.z += v.z;
        return this;
    }

    public subtract(v: Vector3): Vector3 {
        this.x -= v.x;
        this.y -= v.y;
        this.z -= v.z;
        return this;
    }

    public multiply(v: Vector3): Vector3 {
        this.x *= v.x;
        this.y *= v.y;
        this.z *= v.z;
        return this;
    }

    public divide(d: number): Vector3 {
        this.x /= d;
        this.y /= d;
        this.z /= d;
        return this;
    }

    public multiplyScalar(s: number): Vector3 {
        this.x *= s;
        this.y *= s;
        this.z *= s;
        return this;
    }

    public negate(): Vector3 {
        return this.multiplyScalar(-1);
    }

    public normalize(): Vector3 {
        var length = this.magnitude;
        if (length === 0)
            return new Vector3();
        return this.divide(length);
    }

    public dot(v: Vector3): number {
        return Vector3.dot(this, v);
    }

    public cross(v: Vector3): Vector3 {
        return Vector3.cross(this, v);
    }

    public get magnitude(): number {
        return Math.sqrt(Vector3.dot(this, this));
    }

    public get sqrMagnitude(): number {
        return Vector3.dot(this, this);
    }

    /**
     * 获取向量的指定分量
     * @param index 分量索引 (0=x, 1=y, 2=z)
     * @returns 对应分量的值
     */
    public getComponent(index: number): number {
        switch (index) {
            case 0: return this.x;
            case 1: return this.y;
            case 2: return this.z;
            default:
                throw new Error(`Invalid component index: ${index}. Must be 0, 1, or 2.`);
        }
    }

    /**
     * 设置向量的指定分量
     * @param index 分量索引 (0=x, 1=y, 2=z)
     * @param value 要设置的值
     */
    public setComponent(index: number, value: number): void {
        switch (index) {
            case 0: this.x = value; break;
            case 1: this.y = value; break;
            case 2: this.z = value; break;
            default:
                throw new Error(`Invalid component index: ${index}. Must be 0, 1, or 2.`);
        }
    }

    /*
     ADDITIONAL FUNCTIONS
     */

    public clone(): Vector3 {
        return new Vector3(this.x, this.y, this.z);
    }

    public equals(v: Vector3): boolean {
        return v.x == this.x && v.y == this.y && v.z == this.z;
    }

    public toString(): string {
        return "[" + this.x + ", " + this.y + ", " + this.z + "]";
    }

    /*
     STATIC FUNCTIONS
     */

    public static add(v1: Vector3, v2: Vector3): Vector3 {
        return new Vector3(v1.x + v2.x, v1.y + v2.y, v1.z + v2.z);
    }

    public static subtract(v1: Vector3, v2: Vector3): Vector3 {
        return new Vector3(v1.x - v2.x, v1.y - v2.y, v1.z - v2.z);
    }

    public static multiply(v1: Vector3, v2: Vector3): Vector3 {
        return new Vector3(v1.x * v2.x, v1.y * v2.y, v1.z * v2.z);
    }

    public static divide(v1: Vector3, v2: Vector3): Vector3 {
        return new Vector3(v1.x / v2.x, v1.y / v2.y, v1.z / v2.z);
    }

    public static multiplyScalar(v1: Vector3, s: number): Vector3 {
        return new Vector3(v1.x * s, v1.y * s, v1.z * s);
    }

    public static lerp(v1: Vector3, v2: Vector3, t: number): Vector3 {
        var v = new Vector3();
        v.x = v1.x + t * (v2.x - v1.x);
        v.y = v1.y + t * (v2.y - v1.y);
        v.z = v1.z + t * (v2.z - v1.z);
        return v;
    }

    public static reflect(v: Vector3, n: Vector3): Vector3 {
        return Vector3.subtract(v, Vector3.multiplyScalar(n, 2 * Vector3.dot(v, n)));
    }

    public static dot(v1: Vector3, v2: Vector3): number {
        return (v1.x * v2.x + v1.y * v2.y + v1.z * v2.z);
    }

    public static cross(v1: Vector3, v2: Vector3): Vector3 {
        var x = v1.y * v2.z - v1.z * v2.y;
        var y = v1.z * v2.x - v1.x * v2.z;
        var z = v1.x * v2.y - v1.y * v2.x;
        return new Vector3(x, y, z);
    }

    public static distance(v1: Vector3, v2: Vector3): number {
        var x = v2.x - v1.x;
        var y = v2.y - v1.y;
        var z = v2.z - v1.z;
        return Math.sqrt(x * x + y * y + z * z);
    }

    public static difference(v1: Vector3, v2: Vector3): Vector3 {
        var dest = new Vector3();

        dest.x = v1.x - v2.x
        dest.y = v1.y - v2.y
        dest.z = v1.z - v2.z

        return dest
    }

    public static angle(v1: Vector3, v2: Vector3): number {
        return Math.acos(Vector3.dot(v1, v2) / (v1.magnitude * v2.magnitude));
    }

    public static normalize(v: Vector3): Vector3 {
        var length = v.magnitude;
        if (length === 0)
            return new Vector3();
        return Vector3.multiplyScalar(v, 1 / length);
    }

    /**
     * 返回两个向量中每对分量的较小值组成的新向量[1,2](@ref)
     * @param a 第一个向量
     * @param b 第二个向量
     * @returns 由各分量较小值组成的新向量
     */
    public static min(a: Vector3, b: Vector3): Vector3 {
        return new Vector3(
            Math.min(a.x, b.x),
            Math.min(a.y, b.y),
            Math.min(a.z, b.z)
        );
    }

    /**
     * 返回两个向量中每对分量的较大值组成的新向量[3,4](@ref)
     * @param a 第一个向量
     * @param b 第二个向量
     * @returns 由各分量较大值组成的新向量
     */
    public static max(a: Vector3, b: Vector3): Vector3 {
        return new Vector3(
            Math.max(a.x, b.x),
            Math.max(a.y, b.y),
            Math.max(a.z, b.z)
        );
    }

    /*
     STATIC VARIABLES
     */

    public static get ZERO() {
        return new Vector3(0, 0, 0);
    }

    public static get ONE() {
        return new Vector3(1, 1, 1);
    }

    public static get RIGHT() {
        return new Vector3(1, 0, 0);
    }

    public static get LEFT() {
        return new Vector3(-1, 0, 0);
    }

    public static get UP() {
        return new Vector3(0, 1, 0);
    }

    public static get DOWN() {
        return new Vector3(0, -1, 0);
    }

    public static get FORWARD() {
        return new Vector3(0, 0, 1);
    }

    public static get BACK() {
        return new Vector3(0, 0, -1);
    }
}

File: cpp/Vector3.cpp
#include "Vector3.h"
#include <sstream>

// 构造函数实现
Vector3::Vector3() : x(0), y(0), z(0) {
    simd = _mm_set_ps(0, z, y, x);  // SIMD 初始化（x,y,z,0）
}

Vector3::Vector3(float x, float y, float z) : x(x), y(y), z(z) {
    simd = _mm_set_ps(0, z, y, x);  // 低3位存储 x,y,z（_mm_set_ps 顺序为 w,z,y,x）
}

Vector3::Vector3(const Vector2& v) : x(v.x), y(v.y), z(0) {
    simd = _mm_set_ps(0, z, y, x);
}

Vector3::Vector3(const Vector4& v) : x(v.x), y(v.y), z(v.z) {
    simd = _mm_set_ps(0, z, y, x);
}

// 成员方法：修改当前向量
Vector3& Vector3::add(const Vector3& v) {
    simd = _mm_add_ps(simd, v.simd);  // SIMD 并行加法
    x = _mm_cvtss_f32(simd);          // 从 SIMD 寄存器更新 x（最低位）
    y = _mm_cvtss_f32(_mm_shuffle_ps(simd, simd, _MM_SHUFFLE(1,1,1,1)));  // 提取 y
    z = _mm_cvtss_f32(_mm_shuffle_ps(simd, simd, _MM_SHUFFLE(2,2,2,2)));  // 提取 z
    return *this;
}

Vector3& Vector3::subtract(const Vector3& v) {
    simd = _mm_sub_ps(simd, v.simd);  // SIMD 并行减法
    x = _mm_cvtss_f32(simd);
    y = _mm_cvtss_f32(_mm_shuffle_ps(simd, simd, _MM_SHUFFLE(1,1,1,1)));
    z = _mm_cvtss_f32(_mm_shuffle_ps(simd, simd, _MM_SHUFFLE(2,2,2,2)));
    return *this;
}

Vector3& Vector3::multiply(const Vector3& v) {
    simd = _mm_mul_ps(simd, v.simd);  // SIMD 并行乘法
    x = _mm_cvtss_f32(simd);
    y = _mm_cvtss_f32(_mm_shuffle_ps(simd, simd, _MM_SHUFFLE(1,1,1,1)));
    z = _mm_cvtss_f32(_mm_shuffle_ps(simd, simd, _MM_SHUFFLE(2,2,2,2)));
    return *this;
}

Vector3& Vector3::divide(float d) {
    __m128 scalar = _mm_set1_ps(d);   // 将标量扩展为 SIMD 向量（d,d,d,d）
    simd = _mm_div_ps(simd, scalar);  // 并行除法
    x = _mm_cvtss_f32(simd);
    y = _mm_cvtss_f32(_mm_shuffle_ps(simd, simd, _MM_SHUFFLE(1,1,1,1)));
    z = _mm_cvtss_f32(_mm_shuffle_ps(simd, simd, _MM_SHUFFLE(2,2,2,2)));
    return *this;
}

Vector3& Vector3::multiplyScalar(float s) {
    __m128 scalar = _mm_set1_ps(s);   // 标量扩展为 SIMD 向量
    simd = _mm_mul_ps(simd, scalar);  // 并行乘法
    x = _mm_cvtss_f32(simd);
    y = _mm_cvtss_f32(_mm_shuffle_ps(simd, simd, _MM_SHUFFLE(1,1,1,1)));
    z = _mm_cvtss_f32(_mm_shuffle_ps(simd, simd, _MM_SHUFFLE(2,2,2,2)));
    return *this;
}

Vector3& Vector3::negate() {
    return multiplyScalar(-1.0f);
}

Vector3& Vector3::normalize() {
    float len = magnitude();
    if (len == 0) {
        x = y = z = 0;
        simd = _mm_set_ps(0, 0, 0, 0);
        return *this;
    }
    return divide(len);
}

// 成员方法：计算结果
float Vector3::dot(const Vector3& v) const {
    return Vector3::dot(*this, v);
}

Vector3 Vector3::cross(const Vector3& v) const {
    return Vector3::cross(*this, v);
}

float Vector3::magnitude() const {
    return std::sqrt(sqrMagnitude());
}

float Vector3::sqrMagnitude() const {
    return Vector3::dot(*this, *this);
}

// 分量访问
float Vector3::getComponent(int index) const {
    switch (index) {
        case 0: return x;
        case 1: return y;
        case 2: return z;
        default: throw std::invalid_argument("Invalid component index (must be 0,1,2)");
    }
}

void Vector3::setComponent(int index, float value) {
    switch (index) {
        case 0: x = value; break;
        case 1: y = value; break;
        case 2: z = value; break;
        default: throw std::invalid_argument("Invalid component index (must be 0,1,2)");
    }
    simd = _mm_set_ps(0, z, y, x);  // 更新 SIMD 寄存器
}

// 工具方法
Vector3 Vector3::clone() const {
    return Vector3(x, y, z);
}

bool Vector3::equals(const Vector3& v) const {
    return x == v.x && y == v.y && z == v.z;
}

std::string Vector3::toString() const {
    std::stringstream ss;
    ss << "[" << x << ", " << y << ", " << z << "]";
    return ss.str();
}

// 静态方法：向量运算
Vector3 Vector3::add(const Vector3& v1, const Vector3& v2) {
    __m128 res = _mm_add_ps(v1.simd, v2.simd);
    return Vector3(
        _mm_cvtss_f32(res),
        _mm_cvtss_f32(_mm_shuffle_ps(res, res, _MM_SHUFFLE(1,1,1,1))),
        _mm_cvtss_f32(_mm_shuffle_ps(res, res, _MM_SHUFFLE(2,2,2,2)))
    );
}

Vector3 Vector3::subtract(const Vector3& v1, const Vector3& v2) {
    __m128 res = _mm_sub_ps(v1.simd, v2.simd);
    return Vector3(
        _mm_cvtss_f32(res),
        _mm_cvtss_f32(_mm_shuffle_ps(res, res, _MM_SHUFFLE(1,1,1,1))),
        _mm_cvtss_f32(_mm_shuffle_ps(res, res, _MM_SHUFFLE(2,2,2,2)))
    );
}

Vector3 Vector3::multiply(const Vector3& v1, const Vector3& v2) {
    __m128 res = _mm_mul_ps(v1.simd, v2.simd);
    return Vector3(
        _mm_cvtss_f32(res),
        _mm_cvtss_f32(_mm_shuffle_ps(res, res, _MM_SHUFFLE(1,1,1,1))),
        _mm_cvtss_f32(_mm_shuffle_ps(res, res, _MM_SHUFFLE(2,2,2,2)))
    );
}

Vector3 Vector3::divide(const Vector3& v1, const Vector3& v2) {
    __m128 res = _mm_div_ps(v1.simd, v2.simd);
    return Vector3(
        _mm_cvtss_f32(res),
        _mm_cvtss_f32(_mm_shuffle_ps(res, res, _MM_SHUFFLE(1,1,1,1))),
        _mm_cvtss_f32(_mm_shuffle_ps(res, res, _MM_SHUFFLE(2,2,2,2)))
    );
}

Vector3 Vector3::multiplyScalar(const Vector3& v1, float s) {
    __m128 scalar = _mm_set1_ps(s);
    __m128 res = _mm_mul_ps(v1.simd, scalar);
    return Vector3(
        _mm_cvtss_f32(res),
        _mm_cvtss_f32(_mm_shuffle_ps(res, res, _MM_SHUFFLE(1,1,1,1))),
        _mm_cvtss_f32(_mm_shuffle_ps(res, res, _MM_SHUFFLE(2,2,2,2)))
    );
}

Vector3 Vector3::lerp(const Vector3& v1, const Vector3& v2, float t) {
    return Vector3(
        v1.x + t * (v2.x - v1.x),
        v1.y + t * (v2.y - v1.y),
        v1.z + t * (v2.z - v1.z)
    );
}

Vector3 Vector3::reflect(const Vector3& v, const Vector3& n) {
    return subtract(v, multiplyScalar(n, 2 * dot(v, n)));
}

float Vector3::dot(const Vector3& v1, const Vector3& v2) {
    __m128 mul = _mm_mul_ps(v1.simd, v2.simd);  // (x1x2, y1y2, z1z2, 0)
    // 提取 x1x2 + y1y2 + z1z2（通过水平加法）
    __m128 shuf = _mm_shuffle_ps(mul, mul, _MM_SHUFFLE(2, 3, 0, 1));
    __m128 sum = _mm_add_ss(mul, shuf);
    shuf = _mm_shuffle_ps(sum, sum, _MM_SHUFFLE(1, 0, 3, 2));
    sum = _mm_add_ss(sum, shuf);
    return _mm_cvtss_f32(sum);
}

Vector3 Vector3::cross(const Vector3& v1, const Vector3& v2) {
    // 叉积公式：(y1z2 - z1y2, z1x2 - x1z2, x1y2 - y1x2)
    return Vector3(
        v1.y * v2.z - v1.z * v2.y,
        v1.z * v2.x - v1.x * v2.z,
        v1.x * v2.y - v1.y * v2.x
    );
}

float Vector3::distance(const Vector3& v1, const Vector3& v2) {
    return subtract(v1, v2).magnitude();
}

Vector3 Vector3::difference(const Vector3& v1, const Vector3& v2) {
    return subtract(v1, v2);
}

float Vector3::angle(const Vector3& v1, const Vector3& v2) {
    float dotProd = dot(v1, v2);
    float magProduct = v1.magnitude() * v2.magnitude();
    return std::acos(dotProd / magProduct);
}

Vector3 Vector3::normalize(const Vector3& v) {
    float len = v.magnitude();
    if (len == 0) return Vector3();
    return multiplyScalar(v, 1.0f / len);
}

Vector3 Vector3::min(const Vector3& a, const Vector3& b) {
    return Vector3(
        std::min(a.x, b.x),
        std::min(a.y, b.y),
        std::min(a.z, b.z)
    );
}

Vector3 Vector3::max(const Vector3& a, const Vector3& b) {
    return Vector3(
        std::max(a.x, b.x),
        std::max(a.y, b.y),
        std::max(a.z, b.z)
    );
}

// 静态常量定义
const Vector3 Vector3::ZERO = Vector3(0, 0, 0);
const Vector3 Vector3::ONE = Vector3(1, 1, 1);
const Vector3 Vector3::RIGHT = Vector3(1, 0, 0);
const Vector3 Vector3::LEFT = Vector3(-1, 0, 0);
const Vector3 Vector3::UP = Vector3(0, 1, 0);
const Vector3 Vector3::DOWN = Vector3(0, -1, 0);
const Vector3 Vector3::FORWARD = Vector3(0, 0, 1);
const Vector3 Vector3::BACK = Vector3(0, 0, -1);